File: kirun-js/src/engine/runtime/expression/ExpressionLexer.ts
import { ExpressionEvaluationException } from './exception/ExpressionEvaluationException';

export enum TokenType {
    IDENTIFIER = 'IDENTIFIER',
    NUMBER = 'NUMBER',
    STRING = 'STRING',
    OPERATOR = 'OPERATOR',
    LEFT_PAREN = 'LEFT_PAREN',
    RIGHT_PAREN = 'RIGHT_PAREN',
    LEFT_BRACKET = 'LEFT_BRACKET',
    RIGHT_BRACKET = 'RIGHT_BRACKET',
    DOT = 'DOT',
    QUESTION = 'QUESTION',
    COLON = 'COLON',
    WHITESPACE = 'WHITESPACE',
    EOF = 'EOF',
}

export class Token {
    constructor(
        public type: TokenType,
        public value: string,
        public startPos: number,
        public endPos: number,
    ) {}

    public toString(): string {
        return `Token(${this.type}, "${this.value}", ${this.startPos}-${this.endPos})`;
    }
}

export class ExpressionLexer {
    private pos: number = 0;
    private readonly input: string;
    private readonly length: number;

    constructor(expression: string) {
        this.input = expression;
        this.length = expression.length;
    }

    public getPosition(): number {
        return this.pos;
    }

    public peek(): Token | null {
        const savedPos = this.pos;
        const token = this.nextToken();
        this.pos = savedPos;
        return token;
    }

    public nextToken(): Token | null {
        if (this.pos >= this.length) {
            return new Token(TokenType.EOF, '', this.pos, this.pos);
        }

        // Skip whitespace
        while (this.pos < this.length && this.isWhitespace(this.input[this.pos])) {
            this.pos++;
        }

        if (this.pos >= this.length) {
            return new Token(TokenType.EOF, '', this.pos, this.pos);
        }

        const startPos = this.pos;
        const char = this.input[this.pos];

        // Handle string literals
        if (char === '"' || char === "'") {
            return this.readStringLiteral(char, startPos);
        }

        // Handle numbers
        // Note: Only treat '-' as negative sign at start or after operator/open paren, not after identifiers/numbers
        if (this.isDigit(char)) {
            return this.readNumber(startPos);
        }

        // Handle operators and special characters
        switch (char) {
            case '(':
                this.pos++;
                return new Token(TokenType.LEFT_PAREN, '(', startPos, this.pos);
            case ')':
                this.pos++;
                return new Token(TokenType.RIGHT_PAREN, ')', startPos, this.pos);
            case '[':
                this.pos++;
                return new Token(TokenType.LEFT_BRACKET, '[', startPos, this.pos);
            case ']':
                this.pos++;
                return new Token(TokenType.RIGHT_BRACKET, ']', startPos, this.pos);
            case '.':
                // Check for '..' range operator
                if (this.pos + 1 < this.length && this.input[this.pos + 1] === '.') {
                    this.pos += 2;
                    return new Token(TokenType.OPERATOR, '..', startPos, this.pos);
                }
                this.pos++;
                return new Token(TokenType.DOT, '.', startPos, this.pos);
            case '?':
                this.pos++;
                return new Token(TokenType.QUESTION, '?', startPos, this.pos);
            case ':':
                this.pos++;
                return new Token(TokenType.COLON, ':', startPos, this.pos);
        }

        // Handle multi-character operators
        const operator = this.readOperator(startPos);
        if (operator) {
            return operator;
        }

        // Handle identifiers and keywords
        return this.readIdentifier(startPos);
    }

    private readStringLiteral(quoteChar: string, startPos: number): Token {
        this.pos++; // Skip opening quote
        let value = '';
        let escaped = false;

        while (this.pos < this.length) {
            const char = this.input[this.pos];
            
            if (escaped) {
                value += char;
                escaped = false;
                this.pos++;
                continue;
            }

            if (char === '\\') {
                escaped = true;
                value += char;
                this.pos++;
                continue;
            }

            if (char === quoteChar) {
                this.pos++; // Skip closing quote
                const endPos = this.pos;
                // Return the full string including quotes for bracket notation detection
                return new Token(TokenType.STRING, quoteChar + value + quoteChar, startPos, endPos);
            }

            value += char;
            this.pos++;
        }

        throw new ExpressionEvaluationException(
            this.input,
            `Missing string ending marker ${quoteChar}`,
        );
    }

    private readNumber(startPos: number): Token {
        let value = '';
        let hasDecimal = false;

        // Note: Negative sign is handled by the parser as unary minus, not by the lexer
        // This avoids the ambiguity between "a-1" (subtraction) and "-1" (negative number)

        while (this.pos < this.length) {
            const char = this.input[this.pos];
            
            if (this.isDigit(char)) {
                value += char;
                this.pos++;
            } else if (char === '.' && !hasDecimal && this.pos + 1 < this.length && this.isDigit(this.input[this.pos + 1])) {
                value += char;
                hasDecimal = true;
                this.pos++;
            } else {
                break;
            }
        }

        return new Token(TokenType.NUMBER, value, startPos, this.pos);
    }

    private readOperator(startPos: number): Token | null {
        // Try longest operators first
        const operators = [
            '>>>', '<<', '>>', '<=', '>=', '!=', '==', '//',
            '+', '-', '*', '/', '%', '=', '<', '>', '&', '|', '^', '~',
        ];

        for (const op of operators) {
            if (this.pos + op.length <= this.length) {
                const candidate = this.input.substring(this.pos, this.pos + op.length);
                if (candidate === op) {
                    this.pos += op.length;
                    return new Token(TokenType.OPERATOR, op, startPos, this.pos);
                }
            }
        }

        return null;
    }

    private readIdentifier(startPos: number): Token {
        let value = '';

        while (this.pos < this.length) {
            const char = this.input[this.pos];
            
            // Handle bracket notation: only include STATIC bracket content in identifier
            // Static: [9], ["key"], ['key'] - numeric literals or quoted strings
            // Dynamic: [Page.id], [expr + 1] - identifiers or expressions
            if (char === '[') {
                // Check if bracket content is static (numeric or quoted string)
                if (this.isStaticBracketContent()) {
                    // Include static bracket content in the identifier
                    value += this.readStaticBracketContent();
                    // Continue reading more of the identifier
                    continue;
                } else {
                    // Dynamic bracket content - stop identifier here
                    // The bracket will be tokenized separately as LEFT_BRACKET
                    break;
                }
            }
            
            // Stop at non-identifier characters
            // Dots are NOT included - they are handled as DOT tokens for OBJECT_OPERATOR
            if (!this.isIdentifierChar(char)) {
                break;
            }
            
            value += char;
            this.pos++;
        }

        if (value.length === 0) {
            throw new ExpressionEvaluationException(
                this.input,
                `Unexpected character: ${this.input[this.pos]}`,
            );
        }

        // Check for keywords that are operators (standalone, not part of identifier)
        const keywordOperators = ['and', 'or', 'not'];
        if (keywordOperators.includes(value.toLowerCase())) {
            return new Token(TokenType.OPERATOR, value.toLowerCase(), startPos, this.pos);
        }

        return new Token(TokenType.IDENTIFIER, value, startPos, this.pos);
    }

    /**
     * Check if the bracket content starting at current position is static.
     * Static content: numeric literals ([9], [-1]) or quoted strings (["key"], ['key'])
     * Dynamic content: identifiers ([Page.id]) or expressions ([expr + 1])
     */
    private isStaticBracketContent(): boolean {
        // Peek ahead to check what's inside the bracket
        let peekPos = this.pos + 1; // Skip the '['
        
        // Skip whitespace
        while (peekPos < this.length && this.isWhitespace(this.input[peekPos])) {
            peekPos++;
        }
        
        if (peekPos >= this.length) return false;
        
        const firstChar = this.input[peekPos];
        
        // Check for quoted string
        if (firstChar === '"' || firstChar === "'") {
            return true;
        }
        
        // Check for numeric literal (including negative numbers)
        if (this.isDigit(firstChar) || (firstChar === '-' && peekPos + 1 < this.length && this.isDigit(this.input[peekPos + 1]))) {
            // Verify it's a pure number followed by ] or ..
            let numEnd = peekPos;
            if (this.input[numEnd] === '-') numEnd++;
            while (numEnd < this.length && (this.isDigit(this.input[numEnd]) || this.input[numEnd] === '.')) {
                // Check for range operator (..)
                if (this.input[numEnd] === '.' && numEnd + 1 < this.length && this.input[numEnd + 1] === '.') {
                    // This is a range operator - still static
                    return true;
                }
                numEnd++;
            }
            // Skip whitespace
            while (numEnd < this.length && this.isWhitespace(this.input[numEnd])) {
                numEnd++;
            }
            // Check if followed by ] or ..
            if (numEnd < this.length && (this.input[numEnd] === ']' || 
                (this.input[numEnd] === '.' && numEnd + 1 < this.length && this.input[numEnd + 1] === '.'))) {
                return true;
            }
            // If followed by anything else (like + or identifier), it's dynamic
            return false;
        }
        
        // Anything else (identifier, operator) is dynamic
        return false;
    }

    /**
     * Read static bracket content and return it as a string (including brackets).
     * Assumes isStaticBracketContent() returned true.
     */
    private readStaticBracketContent(): string {
        let content = '[';
        this.pos++; // Skip the '['
        
        let bracketCount = 1;
        while (this.pos < this.length && bracketCount > 0) {
            const c = this.input[this.pos];
            content += c;
            this.pos++;
            
            if (c === '[') {
                bracketCount++;
            } else if (c === ']') {
                bracketCount--;
            } else if (c === '"' || c === "'") {
                // Handle string literals inside brackets - read until closing quote
                const quoteChar = c;
                while (this.pos < this.length) {
                    const nextChar = this.input[this.pos];
                    content += nextChar;
                    this.pos++;
                    if (nextChar === quoteChar && this.pos > 1 && this.input[this.pos - 2] !== '\\') {
                        break;
                    }
                }
            }
        }
        
        return content;
    }
    
    private isOperatorChar(char: string): boolean {
        // Characters that indicate operators (not part of identifiers)
        return ['+', '-', '*', '/', '%', '=', '!', '<', '>', '&', '|', '?', ':', '(', ')', ';', ','].includes(char);
    }

    private isDigit(char: string): boolean {
        return char >= '0' && char <= '9';
    }

    private isWhitespace(char: string): boolean {
        return char === ' ' || char === '\t' || char === '\n' || char === '\r';
    }

    private isIdentifierChar(char: string): boolean {
        // Identifiers should NOT contain dots - dots are handled by the parser as OBJECT_OPERATOR
        // This preserves the OBJECT_OPERATOR tree structure for evaluation
        return (
            (char >= 'a' && char <= 'z') ||
            (char >= 'A' && char <= 'Z') ||
            (char >= '0' && char <= '9') ||
            char === '_'
        );
    }
}

File: kirun-py/src/kirun_py/runtime/expression/expression_lexer.py
from __future__ import annotations

from enum import Enum
from typing import Optional

from kirun_py.runtime.expression.exception.expression_evaluation_exception import ExpressionEvaluationException


class TokenType(Enum):
    IDENTIFIER = 'IDENTIFIER'
    NUMBER = 'NUMBER'
    STRING = 'STRING'
    OPERATOR = 'OPERATOR'
    LEFT_PAREN = 'LEFT_PAREN'
    RIGHT_PAREN = 'RIGHT_PAREN'
    LEFT_BRACKET = 'LEFT_BRACKET'
    RIGHT_BRACKET = 'RIGHT_BRACKET'
    DOT = 'DOT'
    QUESTION = 'QUESTION'
    COLON = 'COLON'
    WHITESPACE = 'WHITESPACE'
    EOF = 'EOF'


class Token:

    def __init__(self, type_: TokenType, value: str, start_pos: int, end_pos: int):
        self.type = type_
        self.value = value
        self.start_pos = start_pos
        self.end_pos = end_pos

    def __str__(self) -> str:
        return f'Token({self.type.value}, "{self.value}", {self.start_pos}-{self.end_pos})'


class ExpressionLexer:

    def __init__(self, expression: str):
        self._pos: int = 0
        self._input: str = expression
        self._length: int = len(expression)

    def get_position(self) -> int:
        return self._pos

    def peek(self) -> Optional[Token]:
        saved_pos = self._pos
        token = self.next_token()
        self._pos = saved_pos
        return token

    def next_token(self) -> Optional[Token]:
        if self._pos >= self._length:
            return Token(TokenType.EOF, '', self._pos, self._pos)

        # Skip whitespace
        while self._pos < self._length and self._is_whitespace(self._input[self._pos]):
            self._pos += 1

        if self._pos >= self._length:
            return Token(TokenType.EOF, '', self._pos, self._pos)

        start_pos = self._pos
        char = self._input[self._pos]

        # Handle string literals
        if char == '"' or char == "'":
            return self._read_string_literal(char, start_pos)

        # Handle numbers
        if self._is_digit(char):
            return self._read_number(start_pos)

        # Handle operators and special characters
        if char == '(':
            self._pos += 1
            return Token(TokenType.LEFT_PAREN, '(', start_pos, self._pos)
        if char == ')':
            self._pos += 1
            return Token(TokenType.RIGHT_PAREN, ')', start_pos, self._pos)
        if char == '[':
            self._pos += 1
            return Token(TokenType.LEFT_BRACKET, '[', start_pos, self._pos)
        if char == ']':
            self._pos += 1
            return Token(TokenType.RIGHT_BRACKET, ']', start_pos, self._pos)
        if char == '.':
            # Check for '..' range operator
            if self._pos + 1 < self._length and self._input[self._pos + 1] == '.':
                self._pos += 2
                return Token(TokenType.OPERATOR, '..', start_pos, self._pos)
            self._pos += 1
            return Token(TokenType.DOT, '.', start_pos, self._pos)
        if char == '?':
            # Check for ?? (nullish coalescing) before single ? (ternary)
            if self._pos + 1 < self._length and self._input[self._pos + 1] == '?':
                self._pos += 2
                return Token(TokenType.OPERATOR, '??', start_pos, self._pos)
            self._pos += 1
            return Token(TokenType.QUESTION, '?', start_pos, self._pos)
        if char == ':':
            self._pos += 1
            return Token(TokenType.COLON, ':', start_pos, self._pos)

        # Handle multi-character operators
        operator = self._read_operator(start_pos)
        if operator is not None:
            return operator

        # Handle identifiers and keywords
        return self._read_identifier(start_pos)

    def _read_string_literal(self, quote_char: str, start_pos: int) -> Token:
        self._pos += 1  # Skip opening quote
        value = ''
        escaped = False

        while self._pos < self._length:
            char = self._input[self._pos]

            if escaped:
                value += char
                escaped = False
                self._pos += 1
                continue

            if char == '\\':
                escaped = True
                value += char
                self._pos += 1
                continue

            if char == quote_char:
                self._pos += 1  # Skip closing quote
                end_pos = self._pos
                # Return the full string including quotes for bracket notation detection
                return Token(TokenType.STRING, quote_char + value + quote_char, start_pos, end_pos)

            value += char
            self._pos += 1

        raise ExpressionEvaluationException(
            self._input,
            f'Missing string ending marker {quote_char}',
        )

    def _read_number(self, start_pos: int) -> Token:
        value = ''
        has_decimal = False

        while self._pos < self._length:
            char = self._input[self._pos]

            if self._is_digit(char):
                value += char
                self._pos += 1
            elif (char == '.' and not has_decimal and
                  self._pos + 1 < self._length and self._is_digit(self._input[self._pos + 1])):
                value += char
                has_decimal = True
                self._pos += 1
            else:
                break

        return Token(TokenType.NUMBER, value, start_pos, self._pos)

    def _read_operator(self, start_pos: int) -> Optional[Token]:
        # Try longest operators first
        operators = [
            '>>>', '<<', '>>', '<=', '>=', '!=', '==', '//',
            '+', '-', '*', '/', '%', '=', '<', '>', '&', '|', '^', '~',
        ]

        for op in operators:
            if self._pos + len(op) <= self._length:
                candidate = self._input[self._pos:self._pos + len(op)]
                if candidate == op:
                    self._pos += len(op)
                    return Token(TokenType.OPERATOR, op, start_pos, self._pos)

        return None

    def _read_identifier(self, start_pos: int) -> Token:
        value = ''

        while self._pos < self._length:
            char = self._input[self._pos]

            # Handle bracket notation: only include STATIC bracket content in identifier
            if char == '[':
                if self._is_static_bracket_content():
                    value += self._read_static_bracket_content()
                    continue
                else:
                    break

            # Stop at non-identifier characters
            # Dots are NOT included - they are handled as DOT tokens for OBJECT_OPERATOR
            if not self._is_identifier_char(char):
                break

            value += char
            self._pos += 1

        if len(value) == 0:
            raise ExpressionEvaluationException(
                self._input,
                f'Unexpected character: {self._input[self._pos]}',
            )

        # Check for keywords that are operators
        keyword_operators = ['and', 'or', 'not']
        if value.lower() in keyword_operators:
            return Token(TokenType.OPERATOR, value.lower(), start_pos, self._pos)

        return Token(TokenType.IDENTIFIER, value, start_pos, self._pos)

    def _is_static_bracket_content(self) -> bool:
        """Check if the bracket content starting at current position is static.

        Static content: numeric literals ([9], [-1]) or quoted strings (["key"], ['key'])
        Dynamic content: identifiers ([Page.id]) or expressions ([expr + 1])
        """
        peek_pos = self._pos + 1  # Skip the '['

        # Skip whitespace
        while peek_pos < self._length and self._is_whitespace(self._input[peek_pos]):
            peek_pos += 1

        if peek_pos >= self._length:
            return False

        first_char = self._input[peek_pos]

        # Check for quoted string
        if first_char == '"' or first_char == "'":
            return True

        # Check for numeric literal (including negative numbers)
        if self._is_digit(first_char) or (
            first_char == '-' and peek_pos + 1 < self._length and
            self._is_digit(self._input[peek_pos + 1])
        ):
            # Verify it's a pure number followed by ] or ..
            num_end = peek_pos
            if self._input[num_end] == '-':
                num_end += 1
            while num_end < self._length and (
                self._is_digit(self._input[num_end]) or self._input[num_end] == '.'
            ):
                # Check for range operator (..)
                if (self._input[num_end] == '.' and num_end + 1 < self._length and
                        self._input[num_end + 1] == '.'):
                    return True
                num_end += 1
            # Skip whitespace
            while num_end < self._length and self._is_whitespace(self._input[num_end]):
                num_end += 1
            # Check if followed by ] or ..
            if num_end < self._length and (
                self._input[num_end] == ']' or
                (self._input[num_end] == '.' and num_end + 1 < self._length and
                 self._input[num_end + 1] == '.')
            ):
                return True
            return False

        # Anything else is dynamic
        return False

    def _read_static_bracket_content(self) -> str:
        """Read static bracket content and return it as a string (including brackets).

        Assumes _is_static_bracket_content() returned True.
        """
        content = '['
        self._pos += 1  # Skip the '['

        bracket_count = 1
        while self._pos < self._length and bracket_count > 0:
            c = self._input[self._pos]
            content += c
            self._pos += 1

            if c == '[':
                bracket_count += 1
            elif c == ']':
                bracket_count -= 1
            elif c == '"' or c == "'":
                # Handle string literals inside brackets
                quote_char = c
                while self._pos < self._length:
                    next_char = self._input[self._pos]
                    content += next_char
                    self._pos += 1
                    if (next_char == quote_char and self._pos > 1 and
                            self._input[self._pos - 2] != '\\'):
                        break

        return content

    @staticmethod
    def _is_digit(char: str) -> bool:
        return '0' <= char <= '9'

    @staticmethod
    def _is_whitespace(char: str) -> bool:
        return char in (' ', '\t', '\n', '\r')

    @staticmethod
    def _is_identifier_char(char: str) -> bool:
        return (
            ('a' <= char <= 'z') or
            ('A' <= char <= 'Z') or
            ('0' <= char <= '9') or
            char == '_'
        )
